Support recurrent meeting date times

chuabingquan opened the pull request:
## Related issues
Fixes #72.

#### ⚠️ This PR breaks existing data files, you'll probably want to delete `data/app.json` prior to testing.

## Description
<!-- Describe your changes -->
- [x] Use `LocalDateTime` instead of `String` to represent `MeetingDateTime`.
- [x] `MeetingDateTime` accepts and converts a string if it adheres to the `d MMM uuuu h:mma` format (e.g. `10 Feb 2022 3:30pm`).
- [x] Fix test cases and modify storage to support the new date time format.
- [x] Make `MeetingEntry#getDateTime` recurring.
- [x] Update GUI to reflect recurrent datetime change upon the execution of any command.
<!-- If it affects UI, screenshots are highly recommended -->
<img width="852" alt="Screenshot 2022-03-22 at 12 34 28 AM" src="https://user-images.githubusercontent.com/12640352/159309990-05b8b09d-03ab-43e2-bcfe-9abe84a4c5ae.png">

## How to test
<!-- List down the steps to test this, if applicable -->
### 1. Add a meeting with the new datetime format
1. Execute `add n/Tutorial u/https://meet.google.com/newstuff d/22 Mar 2022 12:38am dur/2 m/CS2103 r/Y t/Nope`.
2. You should see a new meeting added with `22 Mar 2022 12:38am` as its date time.

### 2. Test that recurrence is working
Add a new meeting with a recurrence that ends closer to your actual current time. For instance, if it's 10:00am now, and you're creating a meeting with a duration of two hours, then do consider setting the start date and time of the meeting to today, 8:02am. This way, your meeting hasn't ended, and it should still show "Today, 8:02am" in the list. When it's 10:02am, proceed to execute `list`, and the same meeting should now reflect "Today + 7 days, 8:02am".

## Other information
<!--
This section is optional, you may include things like:
- What is left to be done after this
- Any questions or assistance you may require
-->
The current way of getting the GUI to re-render the recurrent date and time is a hack. Effectively, we're forcing a re-render of the entire list every time any command is called. This is done in the interest of time while we explore better and less-complex alternatives on the side. There are currently two ways identified to eventually replace the hack:
1. Perform a check for recurrent updates after the execution of each command, and perform this check on the program's first run too.
2. Perform a re-render whenever the user's mouse (or other events) enters the frame of the GUI. This would make the updates "live", as opposed to waiting for an explicit user action (e.g. executing a command) to trigger the UI changes.

NatalieTanML commented:
Just a suggestion, would it be better to have the user enter the month as a number instead? E.g. `dd MM uuuu ...` It's usually more common to enter it in the numeric format as it would be less likely to have typos, e.g. `June` vs `Jun`

chuabingquan commented:
> > That makes sense actually, I think we can go with that. If so, should we ensure that dashes are required for better readability? (e.g. `d-MM-uuuu ...`)
> > Alternatively, and for the sake of discussion, we could also use [M/L](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) as well to allow users to use one of `7; 07; Jul; July; J`, which would address the problem you brought up as well.
> 
> I like the dashes, but would there be any issues with the updating the parser for the 2nd option?
> 
> Also, I think we can also display the day of the week on the GUI, e.g. `Tue 22 Mar 2022 2:00pm`

I think the second option is trivial to implement, but in hindsight, perhaps we might be confusing our users (e.g. the example usage says `Jun` is expected, but `June` or `07` works). If this is the case, I'll go with your first suggestion and enforce the dashes?

I'll also update the `toString` method to render in the format you've mentioned.

Gnoznaug commented:
I agree with Natalie, I think this format `d-MM-uuuu ...` is good.

chuabingquan commented:
@NatalieTanML @Gnoznaug I've updated to use `d-M-uuuu h:mma`. This way, users don't need to have a leading zero for their month.

NatalieTanML commented on src/main/java/seedu/address/model/meetingentry/MeetingEntry.java:
```diff
@@ -56,7 +58,28 @@ public MeetingUrl getUrl() {
 
     public MeetingDateTime getDateTime() {
-        return dateTime;
+        if (!isRecurring.isRecurring) {
+            return dateTime;
+        }
+        return new MeetingDateTime(getNextRecurrence());
+    }
+
+    private LocalDateTime getNextRecurrence() {
```
Would it be a good idea to include some javadoc for these two methods, as they contain some additional logic under the hood?

NatalieTanML reviewed (changes requested):
Some changes needed wrt exception handling

xMashedxTomatox commented on src/test/java/seedu/address/storage/StorageManagerTest.java:
```diff
@@ -2,7 +2,7 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static seedu.address.testutil.typical.TypicalLinkyTime.getTypicalLinkyTime;
+//import static seedu.address.testutil.typical.TypicalLinkyTime.getTypicalLinkyTime;
```
Remove this line

xMashedxTomatox reviewed (approved):
Functionality wise LGTM, with one minor line removal. Would it be possible for some of the test cases to contain some recurring meetings as well? Locally, I changed some of the test cases to contain recurring meetings and they seem to work just fine.

chuabingquan commented:
> Functionality wise LGTM, with one minor line removal. Would it be possible for some of the test cases to contain some recurring meetings as well? Locally, I changed some of the test cases to contain recurring meetings and they seem to work just fine.

I think the recurring test meetings were made non-recurring after the merge as it was suspected to be causing test case failures. Since the test case failures isn't caused by the recurring test meetings, I've reverted all test meetings to use their original recurrence.

NatalieTanML reviewed (approved):
LGTM, just a small fix regarding the examples

NatalieTanML reviewed (approved):
LGTM
